### PR TITLE
BS-35 | Vijay,Siva | Replace mock response with remote TS server call in SNOMED omod

### PR DESCRIPTION
**Pull request by vijayanandtwks**
Co-authored-by: Siva Reddy <sivareddy.pathuri@thoughtworks.com>

## Requirements

- [x] This PR has a proper title that briefly describes the work done
- [x] I have squashed / amended the comments to make it more redable
- [x] I have included link to all the JIRA ticket('s)
- [x] I wrote the code as a pair or atleast performed a self-review of my own code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Summary
Replace mock response with remote TS server call in SNOMED omod

## JIRA tickets
https://bahmni.atlassian.net/browse/BS-35

### CODE REVIEW

**Comment by vijayanandtwks** [comment]
Incorporated the following Angshu's review comments : 

1)Instead fo GlobalPropertyConstants, move global properties to interface i.e. TerminologyLookupService

2)Move SimpleObjectConstants to VSSimpleObjectMapper 

3)Dont call it ErrorConstants. Call it Errors -> Created Error Enum that has both error code and error description

4)Rename Test Class methods based on "intent" i.e. shouldGetMatchingTerminologies_whenDiagnosisSearchInputParametersPassed_isValid

5)Use Logger Statement to capture specific details regarding errors -> Logger Errors now convey context+error+mitigation  

6)Instead of scenario based errors, Return only standard runtime error and use it in front end to display standard toast error message(handle scenario based errors through logger)

7)VSSimpleObjectMapper uses non-network logic. Create mock from response json to also validate API contract 

8)@Authorized - Reuse "Get Concepts" instead of "Search Terminology Server" as there should not be any difference between Local Concepts vs TS search 

9)Controller - Use Exception instead of TerminologyServicesException

10)Investigate on Apache Jena - It is for test dependency only

File: api/src/main/java/org/bahmni/module/fhirterminologyservices/api/Error.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package org.bahmni.module.fhirterminologyservices.api;
+
+public enum Error {
+    TERMINOLOGY_SERVICES_CONFIG_INVALID("could not connect to terminology server; at least 1 given global property isn't valid i.e. ts.fhir.diagnosissearch.valueseturl, ts.fhir.valueset.urltemplate"),
```
[the remark this answers is not in the document]
Comment (vijayanandtwks) [reply api/src/main/java/org/bahmni/module/fhirterminologyservices/api/Error.java]: The Messages are meant only for logging and they are not used for TerminologyServicesException.

File: api/src/main/java/org/bahmni/module/fhirterminologyservices/api/impl/TerminologyLookupServiceImpl.java [hunk 2]
```diff
@@ -28,32 +35,76 @@ public TerminologyLookupServiceImpl(ValueSetMapper<List<SimpleObject>> vsSimpleO
...
 
     @Override
-    public List<SimpleObject> getResponseList(String searchTerm, Integer limit, String locale) throws IOException {
-        if (MOCK_DIAGNOSES_SEARCH_TERM.contains(searchTerm)) {
-            ValueSet valueSet = createMockFhirTerminologyResponseValueSet();
-            return vsSimpleObjectMapper.map(valueSet);
-        } else {
-            throw new TerminologyServicesException(TERMINOLOGY_SERVER_DOWN_ERROR_MESSAGE);
+    public List<SimpleObject> getResponseList(String searchTerm, Integer limit, String lang) {
+        List<SimpleObject> responseList = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (vijayanandtwks) [reply api/src/main/java/org/bahmni/module/fhirterminologyservices/api/impl/TerminologyLookupServiceImpl.java]: Implemented guard statement as the first statement

File: api/src/main/java/org/bahmni/module/fhirterminologyservices/api/impl/TerminologyLookupServiceImpl.java [hunk 3]
```diff
@@ -28,32 +35,76 @@ public TerminologyLookupServiceImpl(ValueSetMapper<List<SimpleObject>> vsSimpleO
...
+                responseList = vsSimpleObjectMapper.map(valueSet);
+            } catch (Exception exception) {
+                handleException(exception);
+            }
         }
+        return responseList;
+    }
+
+    private ValueSet fetchValueSet(String valueSetEndPoint) {
+        return FhirContext.forR4().newRestfulGenericClient(getTerminologyServerBaseUrl()).read().resource(ValueSet.class).withUrl(valueSetEndPoint).execute();
```
[the remark this answers is not in the document]
Comment (vijayanandtwks) [reply api/src/main/java/org/bahmni/module/fhirterminologyservices/api/impl/TerminologyLookupServiceImpl.java]: Moved FhirContext.forR4() as instance variable

File: api/src/main/java/org/bahmni/module/fhirterminologyservices/api/impl/TerminologyLookupServiceImpl.java [hunk 4]
```diff
@@ -28,32 +35,76 @@ public TerminologyLookupServiceImpl(ValueSetMapper<List<SimpleObject>> vsSimpleO
...
     }
 
+    private String getValueSetEndPoint(String valueSetUrl, String searchTerm, Integer recordLimit, String localeLanguage, boolean includeDesignations) throws UnsupportedEncodingException, TerminologyServicesException {
+        String baseUrl = getTerminologyServerBaseUrl();
```
[the remark this answers is not in the document]
Comment (vijayanandtwks) [reply api/src/main/java/org/bahmni/module/fhirterminologyservices/api/impl/TerminologyLookupServiceImpl.java]: Implemented guard condition as first statement. Also refactored the code using a reusable method "getTSGlobalProperty" that has inbuilt guard statement

**Review by binduak — APPROVED** [review]

**Review by angshu — APPROVED** [review]

**Pull request merged**